### PR TITLE
(fix) if conditions control flow

### PR DESCRIPTION
**Pull request by dummdidumm**
This adds an IfScope to svelte2tsx which keeps track of the if scope including nested ifs. That scope is used to prepend the resulting condition check to all inner lambda functions that svelte2tsx needs to create for slot, each, await. This way, TypeScript's control flow can determine the type of checked variables and no longer loses that info due to the lambda functions.
#619

TODO: 
- ~handle shadowed variables~
- ~remove hidden checks from getDefinitions/findReferences/diagnostics~
- ~fix rename~

### CODE REVIEW

**Comment by dummdidumm** [comment]
@jasonlyu123 could you review this? The logic is a little complicated and I changed code in quite a few places (although most of the new code is tests) and I want to make sure I'm not doing something stupid here or if the approach I took here is not good in your opinion.

High level summary: The if-conditions are repeated and prepended to the lambda functions to reinstantiate control flow. I introduced comments around them to be able to filter those repeats out when returning things like diagnostics,rename etc. That comment could possibly used in other places, too, later on.

**Review by jasonlyu123 — COMMENTED** [review]
Overall great. Didn't found many problems. Just some minor one

File: packages/svelte2tsx/src/htmlxtojsx/nodes/if-scope.ts [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        const ifScope = new IfScope(this.scope, { condition, type: IfType.If }, this);
+        this.child = ifScope;
+    }
+
+    /**
+     * Adds a `else if` branch to the scope and enhances the condition accordingly.
+     */
+    addElseIf(expression: Node, str: MagicString): void {
+        const condition = this.getConditionInfo(str, expression);
+        this.current = {
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/if-scope.ts:213]: Good point!

File: packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            this.value = this.value.parent;
+        }
+    }
+
+    awaitEnter(node: Node) {
+        this.value = this.value.child();
+        if (node.value) {
+            this.handleScope(node.value);
+        }
+        if (node.error) {
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts:55]: What do you mean by "it also appears in the template scope of the slot props resolve"? Could you post a snippet?

[the remark this answers is not in the document]
Comment (jasonlyu123) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts:55]:
```svelte
<script lang="ts">
let hi: string | undefined;
let aPromise;
</script>

  {#await aPromise then hi}
  {:catch}
    <slot prop={hi}">
  {/await}
```

We can also fix this in another PR

File: packages/svelte2tsx/src/htmlxtojsx/nodes/if-scope.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/if-scope.ts]: added a commit with more docs, let me know what you think

File: packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts:55]: This is trickier than I though. We need to do the condition replacements for both then and catch at first but then only use part of it in the then/catch branch, leaving out the replacements for the other half.

[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/nodes/template-scope.ts:55]: Fixed it by refactoring the await block into separate functions. As a nice side effect I think that the handleAwait functions are more understandable now. Let me know what you think 😄

**Review by jasonlyu123 — APPROVED** [review]
Looks good to me 👍

**Pull request merged**